envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt

```diff
@@ -0,0 +1,49 @@
+package pl.allegro.tech.servicemesh.envoycontrol.snapshot.debug
+
+import io.envoyproxy.controlplane.cache.NodeGroup
+import io.envoyproxy.controlplane.cache.Snapshot
+import io.envoyproxy.controlplane.cache.SnapshotCache
+import io.envoyproxy.envoy.api.v2.core.Node
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RestController
+import pl.allegro.tech.servicemesh.envoycontrol.ControlPlane
+import pl.allegro.tech.servicemesh.envoycontrol.groups.Group
+
+@RestController
+class SnapshotDebugController(controlPlane: ControlPlane) {
+    val cache: SnapshotCache<Group> = controlPlane.cache
+    val nodeGroup: NodeGroup<Group> = controlPlane.nodeGroup
+
+    /**
+     * Returns a textual representation of the snapshot for debugging purposes.
+     * It contains the versions of XDS resources and the contents for a provided node JSON
+     * extracted from Envoy's config_dump endpoint.
+     */
+    @PostMapping("/snapshot")
+    fun snapshot(@RequestBody node: Node): ResponseEntity<String> {
+        val nodeHash = nodeGroup.hash(node)
+        val snapshot = cache.getSnapshot(nodeHash)
+        return if (snapshot == null) {
+            return ResponseEntity("snapshot missing", HttpStatus.NOT_FOUND)
+        } else {
+            ResponseEntity(
+                versions(snapshot) +
+                    "snapshot:\n" +
+                    snapshot.toString(),
+                HttpStatus.OK
+            )
+        }
+    }
+
+    private fun versions(snapshot: Snapshot): String {
+        val versions = StringBuilder()
```

Comment:
IMO it is important to return hash in the same format, that is presented in envoy stats (*.version metrics). Only this way this endpoint would be useful verify, if envoy received lastest snapshot.

Right now we show versions in format like this: `endpoints: 9d90d44dbc8c4ff5bfc06f9ec475c062`, but in envoy metrics we have format like this: `version: 23472379237823` (only numbers)

Reply:
A snapshot version provided by an envoy-control can be retrieved from an envoy using `config_dump`.  A version returned in metrics is different and I think we can merge this PR and down the road think of a method of correlation these two versions.

Reply:
I agree we can merge it now, but I think we need to find a way to compare EDS version from this endpoint to Envoy metrics version, because there is no EDS version on config_dump endpoint.